rfcs/0010-multi-tenant-workload-identity/README.md

```diff
@@ -0,0 +1,1270 @@
+allowing permissions to be granted directly to `ServiceAccounts` (if it needs to
+support granting the impersonation permission, then it can probably also easily
+support granting any other permissions depending on the implementation). GCP for
+example has implemented this feature [recently](https://cloud.google.com/blog/products/identity-security/make-iam-for-gke-easier-to-use-with-workload-identity-federation), a GCP IAM
+Service Account is no longer required for workload identity, i.e. GCP IAM
+permissions can now be granted directly to Kubernetes `ServiceAccounts`. This is
+a significant improvement in the user experience, as it significantly reduces
+the required configuration steps. AWS implemented a similar feature called *EKS
+Pod Identity*, but it still requires an IAM Role to be associated with the
+`ServiceAccount`. The minor improvement from the user experience perspective is
+that this association is implemented entirely in the AWS EKS/IAM APIs, no
+annotations are required in the Kubernetes `ServiceAccount`. Another improvement
+from this EKS feature compared to *IAM Roles for Service Accounts* is that users
+no longer need to create an *OIDC Provider* for the EKS cluster in the IAM API.
+
+In sight of the technical background presented above, our proposal becomes simpler.
+The only solution to support multi-tenant workload identity at the object-level for
+the Flux APIs is to associate the Flux objects with Kubernetes `ServiceAccounts`.
+We propose to build the `ServiceAccount` token creation and exchange logic into
+the Flux controllers through a library in the `github.com/fluxcd/pkg` repository.
+
+### API Changes and Controller Flags
+
+For every Flux API that needs to interact with cloud providers, we propose adding
+the field `spec.serviceAccountName` for specifying the Kubernetes `ServiceAccount`
+on the same namespace of the object that must be used for getting access to the
+respective cloud resources. This field would be optional, and when not present the
+original behavior would be observed, i.e. the feature only activates
+when the field is present and a cloud provider among `aws`, `azure` or `gcp` is
+specified in the `spec.provider` field (or similar, see the last paragraph of this
+section). If only the `spec.provider` field is present, then the controller would
+use single-tenant workload identity, i.e. the identity of the controller itself.
+
+Alongside `spec.serviceAccountName` in the Flux APIs, we propose adding the binary
+flag `--use-service-account-for-provider-auth`. By default this flag would be set
+to `false`, meaning that the controller would use single-tenant workload identity
+for all the objects, i.e. the identity of the controller itself. When set to `true`,
+workload identity would require the `spec.serviceAccountName` field to be present
+in the object, and the controller would use the identity of the `ServiceAccount`
+specified in the object for accessing the cloud provider resources.
```

Comment:
This would break how multi-tenancy works now in kustomize-controller and helm-controller which have a `--default-service-account` flag for setting the SA when `spec.serviceAccountName` is not specified. For these two controllers we can't make `spec.serviceAccountName` required, maybe the best way would be to add `--default-service-account` to all SC, NC, IAC, IRC as well.

Reply:
But I see your point, if you enable the flag and specify `spec.decryption.keyProvider` then now you are forced to add a service account and configure RBAC for it. Maybe we should have separated SA fields for `Kustomization` after all.

Reply:
As discussed via DM, we're doing the following changes:

In SC, NC, IRC and IAC we will introduce the opt-in feature gate `EnableProviderServiceAccount`. This gate must be enabled for using MT WI when both `spec.provider` and `spec.serviceAccountName` are specified. In future versions this gate will become opt-out.

In KC we will have `spec.decryption.serviceAccountName` to specify a SA for being used with a cloud provider KMS service.

Reply:
(By the way I just noticed we don't have `spec.serviceAccountName` for specifying an imagePullSecret in the `HelmRepository` API for OCI charts, we only have `spec.secretRef`)

Reply:
Hmm I think we don't need the feature gate. If a user specifies a service account containing image pull secrets in the `OCIRepository` and `ImageRepository` APIs, their intention is clearly to use that secret for authenticating with the container registry, the `spec.serviceAccountName` is not used for anything else at the moment.

Even better, the documentation states that `spec.provider` is specifically for *secret-less* authentication! Image pull secrets from either `spec.secretRef` or from `spec.serviceAccountName` are only used when `spec.provider` is set to `generic` (or left empty, when it defaults to `generic`).

In other words, the value of `spec.provider` fully determines what `spec.serviceAccountName` should be used for! We don't need a feature gate!

Reply:
I updated the RFC with all of this, thanks!